Update Regex for Phone Number

avknaidu opened the pull request:
Issue: #1821 
<!-- Link to relevant issue. All PRs should be asociated with an issue -->

## PR Type
What kind of change does this PR introduce?
<!-- Please uncomment one ore more that apply to this PR -->

<!-- - Bugfix -->
<!-- - Feature -->
<!-- - Code style update (formatting) -->
Refactoring
<!-- - Build or CI related changes -->
<!-- - Documentation content changes -->
<!-- - Sample app changes -->
<!-- - Other... Please describe: -->


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->


## What is the new behavior?
Improves Phone Number Regex Validation.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tested code with current [supported SDKs](../readme.md#supported)
- [ ] Docs have been added/updated which fit [documentation template](https://github.com/Microsoft/UWPCommunityToolkit/blob/master/docs/.template.md). (for bug fixes / features)
- [ ] Sample in sample app has been added / updated (for bug fixes / features)
    - [ ] Icon has been created (if new sample) following the [Thumbnail Style Guide and templates](https://github.com/Microsoft/UWPCommunityToolkit-design-assets)
- [ ] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Header has been added to all new source files (run *build/UpdateHeaders.bat*)
- [x] Contains **NO** breaking changes


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. 
     Please note that breaking changes are likely to be rejected -->


## Other information

kbrons commented:
Looks good, but I think it'd be nice to add to the docs that, if the regex doesn't cover a specific phone format the developer wants to validate, they can set the regex property themselves with a different regex.

skendrot commented on Microsoft.Toolkit/Extensions/StringExtensions.cs:
```diff
@@ -13,7 +13,7 @@ namespace Microsoft.Toolkit.Extensions
     public static class StringExtensions
     {
-        internal const string PhoneNumberRegex = @"^\s*\+?\s*([0-9][\s-]*){9,}$";
+        internal const string PhoneNumberRegex = @"(\([0-9]{3}\) ?)[0-9]{3}-[0-9]{4}|[0-9]{3}-?[0-9]{3}-?[0-9]{4}|([00|+][0-9]{1,2}) ?[0-9]{3}-?[0-9]{3}-?[0-9]{4}";
```
Doesn't work for standard US phone numbers with the country code
`19876543210` 

Readable format: `1-987-654-3210`
I think this should be covered if we're changing the format

avknaidu replied:
![image](https://user-images.githubusercontent.com/6171200/42695770-5090fc28-8684-11e8-8a9c-1f703b769a20.png)

![image](https://user-images.githubusercontent.com/6171200/42695779-5a74ac80-8684-11e8-9eb9-2df25a5a526d.png)

Works in both cases.

skendrot replied:
It returns a valid phone number true, but if you get the match it would be the first nine digits. It also reports that `1970371774445454545454545455454545454454545` is a valid phone number

avknaidu replied:
i see what you mean. Let me take a look.

nmetulev commented:
@avknaidu, 4.0 code freeze is next week, do you think you can make it with this PR?

nmetulev commented:
ping @avknaidu

avknaidu commented:
@nmetulev changes completed. Things took a u turn for me in the past few months. was not able to complete this on time. 

@skendrot please check now. I changed regex now. Updated docs showing formats that are supported.

avknaidu commented:
ping @skendrot

skendrot commented:
With new changes the following is considered to be a valid phone number `123456-555-1234`. Should this be valid?

avknaidu commented:
@skendrot i personally cannot make a judgement if that format is a valid format or not. I came across a bunch of new formats that i never know a phone number can be written in. So i cannot make a judgement on that particular format.

However from the general format of `[CountryCode(Max 3 Digits)][Phone Number]`, i think it is a valid number. Also note, the `IsPhoneNumber` is something which is only based on regex providing sample phone number formats with just a lightweight Extension where in there is a whole [Library](https://github.com/twcclegg/libphonenumber-csharp) doing this, I think this should be a pretty acceptable solution. 

Let me know your thoughts.

skendrot commented:
Sounds good. Looks like there are conflicts so resolve those and we should be good
